File: scripts/convert_to_quiz.py
#!/usr/bin/env python3
"""
Chuyển tài liệu câu hỏi (định dạng chữ ngắn) -> JSON bộ đề cho kahoot-web.

Cách dùng:
    python convert_to_quiz.py de.txt              # in ra màn hình
    python convert_to_quiz.py de.txt de.json      # ghi ra file de.json

ĐỊNH DẠNG NGUỒN (de.txt) — ngắn gọn, dễ gõ tay hoặc nhờ AI viết:

    # Tên bộ đề
    > Mô tả ngắn (không bắt buộc)

    Thủ đô của Việt Nam?
    * Hà Nội
    - TP.HCM
    - Huế
    - Đà Nẵng

    2 + 2 x 2 = ?  [time=15] [points=1200]
    - 8
    * 6
    - 4

Quy tắc:
  #  đầu dòng  -> tên bộ đề
  >  đầu dòng  -> mô tả
  *  (hoặc +)  -> đáp án ĐÚNG
  -             -> đáp án SAI
  Dòng khác (không rỗng) -> là câu hỏi mới
  [time=..] [points=..] trên dòng câu hỏi -> tuỳ chọn, ghi đè mặc định

Đầu ra khớp đúng schema app dùng (title, description, questions[...]).
"""

import json
import re
import sys

# Mặc định — đổi ở đây nếu muốn
DEFAULT_TIME = 20      # giây / câu
DEFAULT_POINTS = 1000  # điểm / câu


def parse(text: str) -> dict:
    title = "Bộ đề không tên"
    description = ""
    questions = []
    cur = None  # câu hỏi đang xây

    def flush():
        if cur is not None:
            questions.append(cur)

    for lineno, raw in enumerate(text.splitlines(), 1):
        line = raw.strip()
        if not line:
            continue

        if line.startswith("#"):
            title = line.lstrip("#").strip() or title
            continue
        if line.startswith(">"):
            description = line.lstrip(">").strip()
            continue

        if line[0] in "*+-":
            if cur is None:
                raise ValueError(f"Dòng {lineno}: đáp án nhưng chưa có câu hỏi.")
            correct = line[0] in "*+"
            atext = line[1:].strip()
            if not atext:
                raise ValueError(f"Dòng {lineno}: đáp án rỗng.")
            cur["answers"].append({"text": atext, "correct": correct})
            continue

        # còn lại: câu hỏi mới
        flush()
        time_v, points_v = DEFAULT_TIME, DEFAULT_POINTS
        m = re.search(r"\[time=(\d+)\]", line)
        if m:
            time_v = int(m.group(1))
        m = re.search(r"\[points=(\d+)\]", line)
        if m:
            points_v = int(m.group(1))
        qtext = re.sub(r"\[(time|points)=\d+\]", "", line).strip()
        cur = {
            "text": qtext,
            "timeLimit": time_v,
            "points": points_v,
            "answers": [],
        }

    flush()

    # kiểm tra hợp lệ
    if not questions:
        raise ValueError("Không tìm thấy câu hỏi nào.")
    for i, q in enumerate(questions, 1):
        if not q["text"]:
            raise ValueError(f"Câu {i}: thiếu nội dung.")
        if len(q["answers"]) < 2:
            raise ValueError(f"Câu {i} ('{q['text'][:30]}'): cần ít nhất 2 đáp án.")
        if not any(a["correct"] for a in q["answers"]):
            raise ValueError(f"Câu {i} ('{q['text'][:30]}'): chưa đánh dấu đáp án đúng (dùng *).")

    return {"title": title, "description": description, "questions": questions}


def main():
    # In tiếng Việt ra console Windows không bị lỗi mã.
    try:
        sys.stdout.reconfigure(encoding="utf-8")
    except Exception:
        pass

    if len(sys.argv) < 2:
        print("Dùng: python convert_to_quiz.py <nguồn.txt> [ra.json]", file=sys.stderr)
        sys.exit(1)

    with open(sys.argv[1], encoding="utf-8") as f:
        quiz = parse(f.read())

    out = json.dumps(quiz, ensure_ascii=False, indent=2)
    if len(sys.argv) >= 3:
        with open(sys.argv[2], "w", encoding="utf-8") as f:
            f.write(out)
        print(f"✓ Đã ghi {sys.argv[2]} — {len(quiz['questions'])} câu.")
    else:
        print(out)


if __name__ == "__main__":
    main()

File: scripts/convert_to_quiz.mjs
// Chuyển tài liệu câu hỏi (định dạng chữ ngắn) -> JSON bộ đề cho kahoot-web.
//
// Cách dùng:
//   node convert_to_quiz.mjs de.txt            # in ra màn hình
//   node convert_to_quiz.mjs de.txt de.json    # ghi ra file de.json
//
// Định dạng nguồn & quy tắc: xem convert_to_quiz.py (giống hệt).

import { readFileSync, writeFileSync } from "fs";

const DEFAULT_TIME = 20; // giây / câu
const DEFAULT_POINTS = 1000; // điểm / câu

function parse(text) {
  let title = "Bộ đề không tên";
  let description = "";
  const questions = [];
  let cur = null;

  const flush = () => {
    if (cur) questions.push(cur);
  };

  const lines = text.split(/\r?\n/);
  lines.forEach((raw, idx) => {
    const lineno = idx + 1;
    const line = raw.trim();
    if (!line) return;

    if (line.startsWith("#")) {
      title = line.replace(/^#+/, "").trim() || title;
      return;
    }
    if (line.startsWith(">")) {
      description = line.replace(/^>+/, "").trim();
      return;
    }

    if ("*+-".includes(line[0])) {
      if (!cur) throw new Error(`Dòng ${lineno}: đáp án nhưng chưa có câu hỏi.`);
      const correct = line[0] === "*" || line[0] === "+";
      const atext = line.slice(1).trim();
      if (!atext) throw new Error(`Dòng ${lineno}: đáp án rỗng.`);
      cur.answers.push({ text: atext, correct });
      return;
    }

    // câu hỏi mới
    flush();
    let time = DEFAULT_TIME;
    let points = DEFAULT_POINTS;
    const t = line.match(/\[time=(\d+)\]/);
    if (t) time = Number(t[1]);
    const p = line.match(/\[points=(\d+)\]/);
    if (p) points = Number(p[1]);
    const qtext = line.replace(/\[(time|points)=\d+\]/g, "").trim();
    cur = { text: qtext, timeLimit: time, points, answers: [] };
  });

  flush();

  if (questions.length === 0) throw new Error("Không tìm thấy câu hỏi nào.");
  questions.forEach((q, i) => {
    const n = i + 1;
    if (!q.text) throw new Error(`Câu ${n}: thiếu nội dung.`);
    if (q.answers.length < 2)
      throw new Error(`Câu ${n} ('${q.text.slice(0, 30)}'): cần ít nhất 2 đáp án.`);
    if (!q.answers.some((a) => a.correct))
      throw new Error(`Câu ${n} ('${q.text.slice(0, 30)}'): chưa đánh dấu đáp án đúng (dùng *).`);
  });

  return { title, description, questions };
}

const [, , src, out] = process.argv;
if (!src) {
  console.error("Dùng: node convert_to_quiz.mjs <nguồn.txt> [ra.json]");
  process.exit(1);
}

const quiz = parse(readFileSync(src, "utf8"));
const json = JSON.stringify(quiz, null, 2);
if (out) {
  writeFileSync(out, json, "utf8");
  console.log(`✓ Đã ghi ${out} — ${quiz.questions.length} câu.`);
} else {
  console.log(json);
}
